### PR TITLE
[13.0][IMP] shopfloor changes in checkout

### PR DESCRIPTION
**Pull request by hparfr**
Add a new mode: "scan & pack" to quickly scan products.

When activated on the stock.picking.type = checkout it will change the view in the mobile app.

Capture d’écran 2021-04-15 à 15 59 42

### CODE REVIEW

**Comment by hparfr** [comment]
@simahawk @sebalix   can you please review this

File: .gitignore [hunk 1]
```diff
@@ -73,3 +73,7 @@ docs/_build/
 
 # OCA rules
 !static/lib/
+
```
Comment (simahawk) [inline .gitignore]: normally I keep these in my global .gitignore

File: shopfloor/actions/data.py [hunk 2]
```diff
@@ -9,6 +9,13 @@
 class DataAction(Component):
     _inherit = "shopfloor.data.action"
 
+    @property
+    def _full_partner_parser(self):
```
Comment (simahawk) [inline shopfloor/actions/data.py]: this is the simpler record parser. Also, "full" should go to the `data_detail` action.

File: shopfloor/actions/data.py [hunk 3]
```diff
@@ -48,6 +55,20 @@ def _picking_parser(self):
             "scheduled_date",
         ]
 
+    @property
+    def _full_picking_parser(self):
```
Comment (simahawk) [inline shopfloor/actions/data.py]: `data_detail` again?

File: shopfloor/actions/data.py [hunk 4]
```diff
@@ -165,10 +186,17 @@ def _move_line_parser(self):
             "id",
             "qty_done",
             "product_uom_qty:quantity",
+            "shopfloor_checkout_done:done",
             ("product_id:product", self._product_parser),
             ("lot_id:lot", self._lot_parser),
             ("location_id:location_src", self._location_parser),
             ("location_dest_id:location_dest", self._location_parser),
+            (
```
Comment (simahawk) [inline shopfloor/actions/data.py]: this will be more expensive than expected. Can't we make it conditional?

File: shopfloor/actions/data.py [hunk 5]
```diff
@@ -243,9 +271,27 @@ def _product_supplier_code(self, rec, field):
...
-        return self._jsonify(record, parser, **kw)
+        if with_pickings == "full":
+            parser.append(("picking_ids:pickings", self._full_picking_parser))
+
+        data = self._jsonify(record, parser, **kw)
+
+        if with_pickings == "full":
+            pickings = record.picking_ids
+
+            for i, picking in enumerate(pickings):
```
Comment (simahawk) [inline shopfloor/actions/data.py]: IMO this should be done in the service where you need it.
Anyway, these move lines computation shall be done in the full_picking_parser.

File: shopfloor/actions/message.py [hunk 6]
```diff
@@ -249,6 +249,15 @@ def x_units_put_in_package(self, qty, product, package):
             ),
         }
 
+    def x_units_put_in_location(self, qty, product, location):
+        # TODO refactor with x_units_put_in_package
```
Comment (simahawk) [inline shopfloor/actions/message.py:268]: maybe just `x_units_put_in(self, qty, product, destination)` and just pick the name of destination.

File: shopfloor/actions/message.py [hunk 7]
```diff
@@ -504,6 +513,12 @@ def packaging_invalid_for_carrier(self, packaging, carrier):
             ),
         }
 
+    def stock_issue_for_line(self, name):
```
Comment (simahawk) [inline shopfloor/actions/message.py]: this is for?

File: shopfloor/actions/schema.py [hunk 8]
```diff
@@ -38,6 +60,12 @@ def move_line(self, with_packaging=False, with_picking=False):
             "package_dest": self._schema_dict_of(
                 self.package(with_packaging=with_packaging), required=False
             ),
+            "suggested_package_dest": self._schema_list_of(
```
Comment (simahawk) [inline shopfloor/actions/schema.py]: why do you need these?

File: shopfloor/services/exception.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
+from functools import wraps
```
Comment (simahawk) [inline shopfloor/services/exception.py:1]: and this is for?

File: shopfloor_mobile/static/wms/src/scenario/checkout.js [hunk 10]
```diff
@@ -320,15 +338,102 @@ const Checkout = {
         return {
             usage: "checkout",
             initial_state_key: "select_document",
+            lastScanned: null,
             states: {
+                scan_products: {
+                    on_scan: scanned => {
```
Comment (simahawk) [inline shopfloor_mobile/static/wms/src/scenario/checkout.js]: way  too much logic here in the definition of the state IMO

File: shopfloor_mobile_base/static/wms/src/components/detail/detail_card.js [hunk 11]
```diff
@@ -58,4 +65,7 @@ Vue.component("item-detail-card", {
         </p>
     </div>
   `,
+    renderTriggered() {
```
Comment (simahawk) [inline shopfloor_mobile_base/static/wms/src/components/detail/detail_card.js]: ?

File: shopfloor_mobile_base/static/wms/src/components/detail/detail_mixin.js [hunk 12]
```diff
@@ -85,14 +85,18 @@ export var ItemDetailMixin = {
             // Defining defaults for an Object property
             // works only if you don't pass the property at all.
             // If you pass only one key, you'll lose all defaults.
-            const opts = _.defaults({}, this.$props.options, {
```
Comment (simahawk) [inline shopfloor_mobile_base/static/wms/src/components/detail/detail_mixin.js:88]: why this change?

**Review by simahawk — CHANGES_REQUESTED** [review]
@hparfr please clean up this PR. Is a bit messy.

1. provide and extended description of the changes and document methods and classes w/ docstrings
2. split mobile changes to their own atomic commit or split them to another PR
3. also, the checkout behavior you need seems more a new checkout scenario: why don't you create a new one? Is it because of the mobile template? If so, we could find a better way to handle it.

File: shopfloor/actions/schema.py (hunk 8 above)
Comment (FranzPoize) [reply shopfloor/actions/schema.py]: We need these when we process order with multiple products. We need to suggest the destination for the following product that go in the same package.

File: shopfloor/services/exception.py (hunk 9 above)
Comment (FranzPoize) [reply shopfloor/services/exception.py:1]: This was an attempt to find a better way to handle common error in the services. wraps is used to wraps the decorated function and allows us to keep the signature and name of the decorated function.

File: shopfloor_mobile_base/static/wms/src/components/detail/detail_mixin.js (hunk 12 above)
Comment (FranzPoize) [reply shopfloor_mobile_base/static/wms/src/components/detail/detail_mixin.js:88]: There's no good reason to use defaults here. And standard javascript has nearly all utility function that underscore/lodash provides. In the end the plan was to remove underscore from the codebase if that's possible and stick to vanilla js for object literal and array manipulation.

**Comment by hparfr** [comment]
Replaced by a v14 branch on akretion/wms.
